### ISSUE TITLE
Hono Client incorrectly handles root endpoints

### ISSUE THREAD

**Issue by sergiocarneiro** [honojs/hono#962]
Defining root endpoints makes the hono client to add a `index` route, which make the requests fail because it doesn't exist.

```ts
const app = new Hono()
  .get('/', (c) => c.text('Pretty Blog API'))

type App = typeof app;

const client = hc<App>("http://localhost:8787/");
```

### Current Behavior
```ts
client.index.$get()
```

### Expected Behavior
```ts
client.$get()
```

**Comment by yusukebe** [comment]
Hi @sergiocarneiro !

It's not documented, but it's expected behavior. A path ending with `/` will have `index` appended to it.

Because we want to write `client.{path}.{method}`, but `client.$get()` will not be `client.{path}.{method}`, so it will be converted.

```ts
client.api.hello.$get() // GOOD: client.{path}.{method}
client.$get() // NOT GOOD: client.{method}
client.index.$get() // GOOD: client.{path}.{method}
```

Instead, when using Hono Client, it is deprecated to use `index` in the path.

**Comment by yudai-nkt** [comment]
While I understand this is intended (after seeing how the types `Client` and `PathToChain` are internally implemented), I also find this behavior a bit less intuitive especially when combined with `Hono.prototype.route`.

Suppose we have a server implementation like 

```ts
import { Hono } from "hono";

const users = new Hono().get("/", (c) =>
  c.jsonT([{ name: "alice" }, { name: "bob" }])
);
const fruits = new Hono().get("/", (c) =>
  c.jsonT([{ name: "apple" }, { name: "berry" }])
);
const app = new Hono().route("/api/users", users).route("/api/fruits", fruits);

export default app;
```

and on client side:

```ts
import { hc } from "hono/client";
import app from "./server";

const { api } = hc<typeof app>("/");
const users = api.users.index.$get();  // I wish this could be just `api.users.$get()`.
const fruits = api.fruits.index.$get();  //and this could be `api.fruits.$get()`.
```

I don't mean to claim this should be fixed and haven't come with up a better idea. That said, IMHO this is worth discussing what we could improve here.

**Comment by sergiocarneiro** [comment]
If `index` is to be kept, then notice that the client is adding it to the path of the requests it sends.

```ts
client.users.index.$get()
``` 
Sends a request to `/users/index` instead of `/users`.

**Comment by yusukebe** [comment]
> this is worth discussing what we could improve here.

OK! I'm going to reconsider it. After that, I'll comment.

**Comment by yudai-nkt** [comment]
To sum up my current opinion; I'm ok with appending `index` to a root path if it is *really* used as a root (it's kinda `/` being mapped to `/index.html`) but would like to avoid it if the Hono instance is `route`d to another Hono instance as shown in my example. The problems are how we can at the time of routing registration know whether a Hono instance is `route`d to another and how hard the implementation might be even if it's possible.

**Comment by yusukebe** [comment]
Hi @sergiocarneiro @yudai-nkt !

This confusion is occurred by the bug or the thing like a bug. This will be fixed at #971 . Can you see it?

Then it works like following. I think these are expected behaviors:

```ts
const users = new Hono().get('/', (c) => c.jsonT([{ name: 'alice' }, { name: 'bob' }]))
const fruits = new Hono().get('/', (c) => c.jsonT([{ name: 'apple' }, { name: 'berry' }]))
const app = new Hono().route('/api/users', users).route('/api/fruits', fruits)

const usersClient = hc<typeof users>('/')
usersClient.index.$get() // OK

const appClient = hc<typeof app>('/')
appClient.api.fruits.$get() // OK
```

**Comment by yudai-nkt** [comment]
I'm afraid that I'll only have mobile internet connection for a while. So I'd be happy if @sergiocarneiro could check/try the PR and could leave some feedback

**Comment by sergiocarneiro** [comment]
Looks good 👍

**Comment by sergiocarneiro** [comment]
Hi @yusukebe,

All good with the **types**, like you mentioned.

However, I also checked whether the requests were being properly **delivered**, and even though I tested it in a complex repository, it's likely issue still remains:
e.g. `usersClient.index.$get()` sends the request to `/index`, instead of `/`, which fails with NOT FOUND.

**Comment by sergiocarneiro** [comment]
Hi @yusukebe,

The types have been fixed 👍

However, I also checked whether the requests were being properly **delivered**, and even though I tested it in a complex repository, it's likely issue still remains:
e.g. `usersClient.index.$get()` sends the request to `/index`, instead of `/`, which fails with NOT FOUND.

**Comment by yusukebe** [comment]
Hi @yudai-nkt @sergiocarneiro 

Thank you!

@sergiocarneiro 

> However, I also checked whether the requests were being properly delivered, and even though I tested it in a complex repository, it's likely issue still remains:

I want to know the entire code or snippets for reproducing the problem. Could you write it down as a comment?

**Comment by sergiocarneiro** [comment]
@yusukebe I'll simply set up a reproduction repo then.

**Comment by sergiocarneiro** [comment]
Hi @yusukebe,

You can use this repository to reproduce the issue:
https://github.com/sergiocarneiro/example-hono-api

### Main Files
1. `packages/api/src/index.ts`
2. `packages/api/src/routes/v1.0/accounts/users.ts`
3. `apps/app/app/routes/_index.tsx`

**Comment by yusukebe** [comment]
@sergiocarneiro 

Thank you! I'll check it later.

**Comment by yusukebe** [comment]
Hi @sergiocarneiro !

Remix! That's a great example!

I've fixed these issues and released "v3.1.0-rc.1". You can try it.

With this release, you will be able to access the path `/v1.0/accounts/users` with `RPC['v1.0'].accounts.users.$get()`.

<img width="501" alt="SS" src="https://user-images.githubusercontent.com/10682/224736639-489fdc77-3900-4dd2-9e32-cf13af6a7494.png">

I think it's working as expected.

**Comment by sergiocarneiro** [comment]
Hi @yusukebe, 
I've tried it and is working 🥳

> Remix! That's a great example!

Thank you! Hope it helps for future end-to-end development
